### PR TITLE
tlvf: add yamls for ap_autoconfig flow

### PR DESCRIPTION
**Pull request by LiorAmram**
-Added yamls:
    eWscAttribute
    eWscValues16
    eWscValues8
    tlvWscM1
    tlvWscM2

   *Based on Wifi Alliance MultiAp specifications:
    https://groups.wi-fi.org/apps/org/workgroup/gmdocs/download.php/91680/Multi-AP_Technical_Specification-180420c.pdf

-Fixed existing yamls:
    tlvApRadioBasicCapabilities - change structure to support nested lists
    tlvApRadioIdentifier - change incorrect struct type used for mac address

-Added new and fixed yamls to tlvf_conf.yaml

Signed-off-by: Lior Amram <lior.amram@intel.com>

### CODE REVIEW

File: tlvf/AutoGenerated/include/tlvf/ieee_1905_1/eWscAttributes.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
+///////////////////////////////////////
```
Comment (tomereli) [inline tlvf/AutoGenerated/include/tlvf/ieee_1905_1/eWscAttributes.h]: Please move the autogenerated files changes to a separate commit

**Review by tomereli — CHANGES_REQUESTED** [review]
- Please move the autogenerated files changes to a separate commit
- Please add more information in the commit message for every added .yaml file
- Please add a comment in the commit message explaining where these definitions came from (the MultiAP specification - but please use the full name or even the download link - https://groups.wi-fi.org/apps/org/workgroup/gmdocs/download.php/91680/Multi-AP_Technical_Specification-180420c.pdf)

File: tlvf/yaml/tlvf/ieee_1905_1/tlvWscM1.yaml [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+  wsc_state:
+    _type: eWscValues8
+    _value_const: WSC_STATE_NOT_CONFIGURED
+
+  manufacturer_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_MANUFACTURER
+  manufacturer_length:
+    _type: uint16_t
+    _value_const: WSC_MAX_NAME_LENGTH
```
Comment (arnout) [inline tlvf/yaml/tlvf/ieee_1905_1/tlvWscM1.yaml]: This can't be right... The string `"foo\0\0\0"` is distinct from the string `"foo"`, as far as I understand, so the length should definitely not always be the maximum length.

AFAIU tlvf should support this kind of construct.

File: tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    _type: eWscAttributes
+    _value_const: ATTR_VENDOR_EXTENSION
+  vendor_extensions_length:
+    _type: uint16_t
+    _value_const: WSC_VENDOR_EXTENSIONS_LENGTH
+  vendor_extensions:
+    _type: uint8_t
+    _length: [ WSC_VENDOR_EXTENSIONS_LENGTH ]
+
+  #encrypted data
```
Comment (arnout) [inline tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml]: The encrypted data should be wrapped in a KEY_WRAP_AUTH attribute. And in hostapd/wpa_supplicant, it is also wrapped in an inner CRED attribute, though the BBF implementation of 1905.1 doesn't do the latter.

File: tlvf/yaml/tlvf/wfa_map/tlvApRadioBasicCapabilities.yaml [hunk 4]
```diff
@@ -9,26 +9,22 @@ tlvApRadioBasicCapabilities:
...
     _type: uint8_t
     _length_var: True
   statically_non_operable_channels_list: 
-    _type: sChannel
-    _length: [ statically_non_operable_channels_list_length ]
-
-sChannel:
-  _type: struct
-  channel: uint8_t
+    _type: uint8_t
```
Comment (arnout) [inline tlvf/yaml/tlvf/wfa_map/tlvApRadioBasicCapabilities.yaml]: I don't understand this change. I guess that means it should be in a separate commit with a proper explanation. And the rest of the changes in this file as well, I think.

**Review by arnout — CHANGES_REQUESTED** [review]

File: tlvf/yaml/tlvf/ieee_1905_1/eWscAttributes.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  ATTR_CONFIG_ERROR: 0x1009
+  ATTR_OS_VERSION: 0x102d
+  ATTR_VENDOR_EXTENSION: 0x1049
+  ATTR_SSID: 0x1045
+  ATTR_AUTH_TYPE: 0x1003
+  ATTR_ENCR_TYPE: 0x100f
+  ATTR_NETWORK_KEY: 0x1027
+  ATTR_KEY_WRAP_AUTH: 0x101e
+  ATTR_ENCR_SETTINGS: 0x1018
+  ATTR_AUTHENTICATOR: 0x1005
```
[the remark this answers is not in the document]
Comment (LiorAmram) [reply tlvf/yaml/tlvf/ieee_1905_1/eWscAttributes.yaml]: I will, but Tomer and I didn't fully understand why it's needed

[the remark this answers is not in the document]
Comment (arnout) [reply tlvf/yaml/tlvf/ieee_1905_1/eWscAttributes.yaml]: It's just whitespace convention, just like the convention not to use \r\n but only \n as newlines.

The only real effect is that if you add a line at the end of the file and there was no newline, you'll get two modified lines in the diff instead of only one.

Also, some editors always add a newline at the end of the file, so if you open a file in such an editor, the last line will be changed even if you have no other modifications.

Finally, when you count the number of lines, some tools will not count that final line without newline as a line, but other tools will. I don't think this is a big deal :-)

[the remark this answers is not in the document]
Comment (LiorAmram) [reply tlvf/yaml/tlvf/ieee_1905_1/eWscAttributes.yaml:39]: Thank you for the explanation :)

File: tlvf/yaml/tlvf/ieee_1905_1/eWscValues16.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  WSC_CONFIG_VIRT_PUSHBUTTON: 0x0280
+  WSC_CONFIG_PHY_PUSHBUTTON: 0x0480
+  DEV_PW_PUSHBUTTON: 0x0004
+  WSC_CFG_NO_ERROR: 0
+  WSC_DEV_NETWORK_INFRA: 6
+  WSC_DEV_NETWORK_INFRA_AP: 1                                                                               
+  WSC_DEV_NETWORK_INFRA_ROUTER: 2                                                                                   
+  WSC_DEV_NETWORK_INFRA_SWITCH: 3                                                                                   
+  WSC_DEV_NETWORK_INFRA_GATEWAY: 4                                                                                    
+  WSC_DEV_NETWORK_INFRA_BRIDGE: 5
```
[the remark this answers is not in the document]
Comment (LiorAmram) [reply tlvf/yaml/tlvf/ieee_1905_1/eWscValues16.yaml:27]: I will refactor in the fixup commit.

File: tlvf/yaml/tlvf/ieee_1905_1/tlvWscM1.yaml [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+  version_attr:
+    _type: eWscAttributes
+    _value_const: ATTR_VERSION
+  version_length:
+    _type: uint16_t
+    _value_const: 1
+  version:
+    _type: eWscValues8
+    _value_const: WSC_VERSION
```
[the remark this answers is not in the document]
Comment (LiorAmram) [reply tlvf/yaml/tlvf/ieee_1905_1/tlvWscM1.yaml:24]: I will refactor in the fixup commit.

File: tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml (hunk 3 above)
Comment (LiorAmram) [reply tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml:232]: Hi, the KEY_WRAP_AUTH should be before the encrypted data or both before and after?
Also, can you attach a reference? because I didn't see KEY_WRAP_AUTH in M2 in the WSC specs, only 'Authenticator' attribute.

Comment (arnout) [reply tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml:232]: I was mistaken, it's wrapped in an ENCRYPTED_SETTINGS attribute, and the last (nested) attribute is KEY_WRAP_AUTH.

So I have four sources for the WSC stuff (and all of them are inconsistent :-)

1. The Multi-AP specification. It doesn't say much about how M1 and M2 have to be constructed, so not much help from there.
2. The WSC specification. This is a huge document and the information is spread all over hte place, so difficult to draw solid conclusions from it. For some reason, the M2 description (v2.0.5, section 8.3.2, p.75) doesn't mention encrypted settings, but later on tables 20 and 21 in section 8.3.9 (M8) do specify that they apply to M2 as well. The "Encrypted Settings" part in section 12 explains how this attribute is constructed, though it is a bit vague as to whether the length is just for the IV or also includes the following (encrypted) attributes. Back to Table 21, it specifies "Credential" as a required attribute, and in section 12, the Credential attribute is specified to contain SSID etc. But now I read the spec again, I see that for the M2 used in AP-Autoconfig, we have to use Table 20 instead of Table 21, and that one doesn't have "Credential", it just has SSID etc. directly there. The description of the Encrypted Settings attribute also specifies that it must end with a Key Wrap Authenticator attribute.
3. The hostapd implementation. Reading the source code, hostapd always assumes that the settings are in the Encrypted Settings attribute, and that the SSID etc. are wrapped into an inner Credential attribute. So it always follows Table 21, AFAICS.
4. The BBF meshComms implementation. This one also has the Encrypted Settings attribute, but doesn't nest into a Credential attribute.
 
Based on all of the above, my conclusion is that we need an Encrypted Settings attribute that contains (cfr. Table 20) type, length (of the entire thing), SSID, AUTH_TYPE, ENCR_TYPE, NETWORK_KEY, BSSID, KEY_WRAP_AUTH. This is exactly how meshComms constructs it. Since hostapd is never involved in this specifc M2, it's not a problem that hostapd does it differently.

Comment (LiorAmram) [reply tlvf/yaml/tlvf/ieee_1905_1/tlvWscM2.yaml:232]: Thank you very much!
The specs are indeed a bit confusing, but you summed it up very well.
I uploaded a PR for the refactoring (in the unified repo) and added this change in a fixup commit.

**Comment by arnout** [comment]
This PR has moved to https://github.com/prplfoundation/prplMesh/pull/9.

All comments have either been fixed there, or have been repeated.
